From 569cd096daa90aa4d4fb8602453843f73feb7fe6 Mon Sep 17 00:00:00 2001
From: Fabian Maurer <dark.shadow4@web.de>
Date: Fri, 20 Jul 2018 21:05:05 +0200
Subject: [PATCH 10/24] wnet: Make WNetGetUniversalNameW return required size
 when buffer is too small and add test

The pointer is set to the required size not only when the input size
is 0, but generally when it is too small.

Signed-off-by: Fabian Maurer <dark.shadow4@web.de>
---
 dlls/mpr/tests/mpr.c | 5 ++++-
 dlls/mpr/wnet.c      | 1 +
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/mpr/tests/mpr.c b/dlls/mpr/tests/mpr.c
index fc067d9..3e5ca09 100644
--- a/dlls/mpr/tests/mpr.c
+++ b/dlls/mpr/tests/mpr.c
@@ -51,11 +51,14 @@ static void test_WNetGetUniversalName(void)
 
         ok(info_size == sizeof(buffer), "Got wrong size: %u\n", info_size);
 
-        fail_size = 0;
+        fail_size = 1;
         ret = WNetGetUniversalNameA(driveA, UNIVERSAL_NAME_INFO_LEVEL,
                 buffer, &fail_size);
         if(drive_type == DRIVE_REMOTE)
+        {
             todo_wine ok(ret == WN_BAD_VALUE || ret == WN_MORE_DATA, "WNetGetUniversalNameA failed: %08x\n", ret);
+            ok(fail_size > 1, "Got %d\n", fail_size);
+        }
         else
             ok(ret == WN_NOT_CONNECTED || ret == WN_NO_NET_OR_BAD_PATH,
                 "(%s) WNetGetUniversalNameW gave wrong error: %u\n", driveA, ret);
diff --git a/dlls/mpr/wnet.c b/dlls/mpr/wnet.c
index ad4f1dd..a2f8a04 100644
--- a/dlls/mpr/wnet.c
+++ b/dlls/mpr/wnet.c
@@ -2385,6 +2385,7 @@ DWORD WINAPI WNetGetUniversalNameW ( LPCWSTR lpLocalPath, DWORD dwInfoLevel,
         size = sizeof(*info) + (lstrlenW(lpLocalPath) + 1) * sizeof(WCHAR);
         if (*lpBufferSize < size)
         {
+            *lpBufferSize = size;
             err = WN_MORE_DATA;
             break;
         }
-- 
1.9.1

